op-supervisor/supervisor/backend/db/sync/client.go

```diff
@@ -0,0 +1,226 @@
+type Client struct {
+	config  Config
+	baseURL string
+
+	httpClient *http.Client
+	retryDelay time.Duration
+}
+
+// NewClient creates a new Client with the given config and server URL.
+func NewClient(config Config, serverURL string, httpClient *http.Client) (*Client, error) {
+	// Verify root directory exists and is actually a directory
+	root, err := filepath.Abs(config.DataDir)
+	if err != nil {
+		return nil, fmt.Errorf("invalid root directory: %w", err)
+	}
+	rootInfo, err := os.Stat(root)
+	if err != nil {
+		return nil, fmt.Errorf("cannot access root directory: %w", err)
+	}
+	if !rootInfo.IsDir() {
+		return nil, fmt.Errorf("root path is not a directory: %s", root)
+	}
+
+	if httpClient == nil {
+		httpClient = DefaultHTTPClient
+	}
+
+	return &Client{
+		config:     config,
+		baseURL:    serverURL,
+		httpClient: httpClient,
+		retryDelay: InitRetryDelay,
+	}, nil
+}
+
+// SyncAll syncs all known databases for the given chains.
+func (c *Client) SyncAll(ctx context.Context, chains []types.ChainID, resume bool) error {
+	for _, chain := range chains {
+		for fileAlias := range Databases {
+			if err := c.SyncDatabase(ctx, chain, fileAlias, resume); err != nil {
+				return fmt.Errorf("failed to sync %s for chain %s: %w", fileAlias, chain, err)
+			}
+		}
+	}
+	return nil
+}
+
+// SyncDatabase downloads the named file from the server.
+// If the local file exists, it will attempt to resume the download if resume is true.
+func (c *Client) SyncDatabase(ctx context.Context, chainID types.ChainID, database Database, resume bool) error {
+	// Validate file alias
+	filePath, exists := Databases[database]
+	if !exists {
+		return fmt.Errorf("unknown file alias: %s", database)
+	}
+
+	// Ensure the chain directory exists
+	chainDir := filepath.Join(c.config.DataDir, chainID.String())
+	if err := os.MkdirAll(chainDir, 0755); err != nil {
+		return fmt.Errorf("failed to create chain directory: %w", err)
+	}
+
+	filePath = filepath.Join(chainDir, filePath)
+
+	// Ensure the database file exists and get initial size
+	filePath = filepath.Join(chainDir, filePath)
+	var initialSize int64
+	if stat, err := os.Stat(filePath); err == nil {
+		initialSize = stat.Size()
+	}
+
+	// If we have data already and don't want to resume then stop now
+	if initialSize > 0 && !resume {
+		return nil
+	}
+
+	// Keep track of the current retry delay
+	currentDelay := c.retryDelay
+	for {
+		// Try to sync and return if successful
+		err := c.attemptSync(ctx, chainID, database, filePath, initialSize)
+		if err == nil {
+			return nil
+		}
+
+		// Check if context was canceled
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
+		c.logError("sync attempt failed", err, database)
+
+		// Wait before retrying
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-time.After(currentDelay):
+			// Double the delay for next time, up to max
+			currentDelay *= 2
+			if currentDelay > MaxRetryDelay {
+				currentDelay = MaxRetryDelay
```

Comment:
You may enjoy using the monorepo's `retry.Do`, which handles this pattern

https://github.com/ethereum-optimism/optimism/blob/f2d5e32bc6cd4ffbe95819b6b5455f7bd6c7454a/op-service/retry/operation.go#L41

Reply:
Hadn't seen this yet, nice 👍